src/main/java/org/opensearch/neuralsearch/processor/TextImageEmbeddingProcessor.java

```diff
@@ -0,0 +1,218 @@
+                }, e -> { handler.accept(null, e); }));
+            }
+        } catch (Exception e) {
+            handler.accept(null, e);
+        }
+
+    }
+
+    private void setVectorFieldsToDocument(IngestDocument ingestDocument, List<Float> vectors) {
+        Objects.requireNonNull(vectors, "embedding failed, inference returns null result!");
+        log.debug("Text embedding result fetched, starting build vector output!");
+        Map<String, Object> textEmbeddingResult = buildTextEmbeddingResult(this.embedding, vectors);
+        textEmbeddingResult.forEach(ingestDocument::setFieldValue);
+    }
+
+    @SuppressWarnings({ "unchecked" })
+    private Map<String, String> createInferences(Map<String, String> knnKeyMap) {
+        Map<String, String> texts = new HashMap<>();
+        if (fieldMap.containsKey(TEXT_FIELD_NAME) && knnKeyMap.containsKey(fieldMap.get(TEXT_FIELD_NAME))) {
+            texts.put(INPUT_TEXT, knnKeyMap.get(fieldMap.get(TEXT_FIELD_NAME)));
+        }
+        if (fieldMap.containsKey(IMAGE_FIELD_NAME) && knnKeyMap.containsKey(fieldMap.get(IMAGE_FIELD_NAME))) {
+            texts.put(INPUT_IMAGE, knnKeyMap.get(fieldMap.get(IMAGE_FIELD_NAME)));
+        }
+        return texts;
+    }
+
+    @VisibleForTesting
+    Map<String, String> buildMapWithKnnKeyAndOriginalValue(IngestDocument ingestDocument) {
+        Map<String, Object> sourceAndMetadataMap = ingestDocument.getSourceAndMetadata();
+        Map<String, String> mapWithKnnKeys = new LinkedHashMap<>();
+        for (Map.Entry<String, String> fieldMapEntry : fieldMap.entrySet()) {
+            String originalKey = fieldMapEntry.getValue(); // field from ingest document that we need to sent as model input, part of
+                                                           // processor definition
+
+            if (!sourceAndMetadataMap.containsKey(originalKey)) {
+                continue;
+            }
+            if (!(sourceAndMetadataMap.get(originalKey) instanceof String)) {
+                throw new IllegalArgumentException("Unsupported format of the field in the document, value must be a string");
+            }
+            mapWithKnnKeys.put(originalKey, (String) sourceAndMetadataMap.get(originalKey));
+        }
+        return mapWithKnnKeys;
+    }
+
+    @SuppressWarnings({ "unchecked" })
+    @VisibleForTesting
+    Map<String, Object> buildTextEmbeddingResult(String knnKey, List<Float> modelTensorList) {
+        Map<String, Object> result = new LinkedHashMap<>();
+        result.put(knnKey, modelTensorList);
+        return result;
+    }
+
+    private void validateEmbeddingFieldsValue(IngestDocument ingestDocument) {
+        Map<String, Object> sourceAndMetadataMap = ingestDocument.getSourceAndMetadata();
+        for (Map.Entry<String, String> embeddingFieldsEntry : fieldMap.entrySet()) {
+            Object sourceValue = sourceAndMetadataMap.get(embeddingFieldsEntry.getKey());
+            if (sourceValue != null) {
```

Comment:
nit. Reduce level of nested block.
```suggestion
            if (sourceValue == null) {
              continue;
            }
```

Reply:
ack